haystack/utils/base_serialization.py

```diff
@@ -69,60 +69,60 @@ def _serialize_value_with_schema(payload: Dict[str, Any]) -> Dict[str, Any]:
     - Primitive types (str, int, float, bool, None)
 
-    :param value: The value to serialize
+    :param payload: The value to serialize (can be any type)
     :returns: The serialized dict representation of the given value. Contains two keys:
-        - "schema": Contains type information for each field
-        - "data": Contains the actual data in a simplified format
+        - "serialization_schema": Contains type information for each field.
+        - "serialized_data": Contains the actual data in a simplified format.
 
     """
-    schema: Dict[str, Any] = {}
-    data: Dict[str, Any] = {}
-
-    for field, val in payload.items():
-        # 1) Handle dataclass‐style objects
-        if hasattr(val, "to_dict") and callable(val.to_dict):
-            type_name = generate_qualified_class_name(type(val))
-            pure = _convert_to_basic_types(val.to_dict())
-            schema[field] = {"type": type_name}
-            data[field] = pure
-
-        # 2) Arbitrary objects w/ __dict__
-        elif hasattr(val, "__dict__"):
-            type_name = generate_qualified_class_name(type(val))
-            pure = _convert_to_basic_types(vars(val))
-            schema[field] = {"type": type_name}
-            data[field] = pure
-
-        # 3) Dicts → "object"
-        elif isinstance(val, dict):
-            pure = _convert_to_basic_types(val)
-            schema[field] = {"type": "object"}
-            data[field] = pure
-
-        # 4) Sequences → "array"
-        elif isinstance(val, (list, tuple, set)):
-            # pure data
-            pure_list = _convert_to_basic_types(list(val))
-            # determine item type from first element (if any)
-            if val:
-                first = next(iter(val))
-                if hasattr(first, "to_dict") and callable(first.to_dict) or hasattr(first, "__dict__"):
-                    item_type = generate_qualified_class_name(type(first))
-                else:
-                    item_type = _primitive_schema_type(first)
-            else:
-                item_type = "any"
-
-            schema[field] = {"type": "array", "items": {"type": item_type}}
-            data[field] = pure_list
-
-        # 5) Primitives
+    # Handle dictionary case - iterate through fields
+    if isinstance(payload, dict):
+        schema: Dict[str, Any] = {}
+        data: Dict[str, Any] = {}
+
+        for field, val in payload.items():
+            # Recursively serialize each field
+            serialized_field = _serialize_value_with_schema(val)
+            schema[field] = serialized_field["serialization_schema"]
+            data[field] = serialized_field["serialized_data"]
+
+        return {"serialization_schema": {"type": "object", "properties": schema}, "serialized_data": data}
+
+    # Handle array case - iterate through elements
+    elif isinstance(payload, (list, tuple, set)):
+        # Convert to list for consistent handling
+        pure_list = _convert_to_basic_types(list(payload))
```

Comment:
By converting to a list we would lose if this was a tuple or a set at deserialization time right? That could be a problem for `set` if the next component expects a set instead of list since they have different member functions e.g. `add()` vs `append()`

Reply:
It seems like we can handle `sets` and `tuples` within the JSON schema spec using:
* for Tuples: `minItems` and `maxItems` (docs [here](https://json-schema.org/understanding-json-schema/reference/array#length))
* for Sets: `uniqueItems` docs [here](https://json-schema.org/understanding-json-schema/reference/array#uniqueItems)